client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -1,182 +1,136 @@
+        def offer1(a: A): Task[Boolean] =
+          enqueue1(a).as(true)
+        def dequeue1: Task[A] = cancellableDequeue1.flatMap { _._1 }
+        def dequeueBatch1(batchSize: Int): Task[NonEmptyChunk[A]] = cancellableDequeueBatch1(batchSize).flatMap { _._1 }
+        def cancellableDequeue1: Task[(Task[A], Task[Unit])] = cancellableDequeueBatch1(1).map { case (deq,cancel) => (deq.map(_.head),cancel) }
+        def cancellableDequeueBatch1(batchSize: Int): Task[(Task[NonEmptyChunk[A]],Task[Unit])] =
+          q.cancellableDequeueBatch1(batchSize).map { case (deq,cancel) => (deq.flatMap(a => permits.incrementBy(a.size.toLong).as(a)), cancel) }
+
+        def size = q.size
+        def full: Signal[Task, Boolean] = q.size.map(_ >= maxTotal)
+        def available: Signal[Task, Int] = q.size.map(maxTotal - _)
+
+        override def build(key: RequestKey): Task[Unit] =
+          permits.tryDecrement.flatMap { b =>
+            if (b) builder(key).flatMap(connection => enqueue1(connection))
+            else q.dequeue1.flatMap(destroy) >> builder(key).flatMap(connection => enqueue1(connection))
           }
-        }
-        else {
-          allocated -= 1
 
-          if (!connection.isClosed) {
-            logger.debug(s"Connection returned was busy.  Shutting down: ${stats}")
-            connection.shutdown()
-          }
+        override def destroy(connection: A): Task[Unit] =
+          connection.isClosed.flatMap{ closed => if (!closed) connection.shutdown() else Task.now(()) }
 
-          if (waitQueue.nonEmpty) {
-            logger.debug(s"Connection returned could not be recycled, new connection needed: ${stats}")
-            val Waiting(key, callback) = waitQueue.dequeue()
-            createConnection(key, callback)
+      }
+    }}
+
+  /**
+    * This method is similar to dequeue except that it works with the key for a Request rather than an element of
+    * the queue directly.
+    * If the buffer is full, we dequeue an element and if it is still open, we either use it if it matches the
+    * RequestKey or we add it back to the queue. If the connection is closed, we add a connection to the pool with the
+    * new RequestKey. If the buffer is not full we create a connection and add it to the queue. This builds the buffer
+    * pool initially.
+    * @param key The Request Key We Need A Connection For
+    * @return A Task of NextConnection which has a connection to use.(NextConnection is wasted here)
+    */
+  def borrow(key: RequestKey): Task[NextConnection] = connectionBuffer.flatMap{q =>
+    q.full.get.flatMap { full =>
+      if (full) {
+        q.dequeue1.flatMap { conn =>
+          conn.isClosed.flatMap { closed =>
+            if (!closed && key == conn.requestKey) {
+              Task.delay(NextConnection(conn, false))
+            }
+            else if (!closed) {
+              q.enqueue1(conn) >>  borrow(key)
```

Comment:
You are correct. This implementation is still very naive in areas. It actually is spinning in the current implementation as my understanding of Semaphore was backward.